File: python/cutfemx/distance.py
# Copyright (c) 2026 ONERA
# Authors: Susanne Claus
# This file is part of CutFEMx
#
# SPDX-License-Identifier:    MIT

from __future__ import annotations

from dataclasses import dataclass
import logging
import time
from typing import Any

import basix.ufl
from dolfinx.fem import Function, functionspace
from dolfinx.mesh import (
    GhostMode,
    Mesh,
    RefinementOption,
    create_cell_partitioner,
    refine,
)

from cutfemx import cutfemx_cpp as _cpp

__all__ = [
    "NormalExtensionResult",
    "SignMode",
    "adapt_mesh_to_stl",
    "build_cell_triangle_map",
    "compute_distance_from_stl",
    "compute_signed_distance",
    "compute_stl_bbox",
    "compute_unsigned_distance",
    "distribute_stl",
    "distribute_stl_facets",
    "from_stl",
    "extend_normal_velocity",
    "reinitialize",
]


SignMode = _cpp.distance.SignMode


@dataclass(frozen=True)
class NormalExtensionResult:
    """Background fields produced by normal-direction velocity extension."""

    speed: Function
    velocity: Function
    signed_distance: Function


def _element_degree(V: Any) -> int:
    degree = V.ufl_element().degree
    return int(degree() if callable(degree) else degree)


def _vector_lagrange_space(mesh: Mesh, degree: int):
    element = basix.ufl.element(
        "Lagrange",
        mesh.basix_cell(),
        degree,
        shape=(mesh.geometry.dim,),
        dtype=mesh.geometry.x.dtype,
    )
    return functionspace(mesh, element)


def compute_stl_bbox(path: str):
    """Compute the STL axis-aligned bounding box without distributing facets."""
    return _cpp.distance.compute_stl_bbox(path)


def distribute_stl(mesh: Mesh, stl_path: str, aabb_padding: float = 0.05):
    """Read and distribute STL facets across the mesh communicator."""
    return _cpp.distance.distribute_stl(
        mesh.comm, mesh._cpp_object, stl_path, aabb_padding
    )


def build_cell_triangle_map(mesh: Mesh, soup, aabb_padding: float = 0.0):
    """Build the local cell-to-STL-triangle map used for near-field distances."""
    return _cpp.distance.build_cell_triangle_map(
        mesh._cpp_object, soup, aabb_padding
    )


def compute_unsigned_distance(
    dist_func: Function,
    soup,
    cell_map,
    max_iter: int = 1000,
    tol: float = 1.0e-10,
) -> None:
    """Fill ``dist_func`` with unsigned surface distance values."""
    _cpp.distance.compute_unsigned_distance(
        dist_func._cpp_object, soup, cell_map, max_iter, tol
    )


def compute_signed_distance(
    dist_func: Function,
    soup,
    cell_map,
    sign_mode: SignMode = SignMode.ComponentAnchor,
    max_iter: int = 1000,
    tol: float = 1.0e-10,
) -> None:
    """Fill ``dist_func`` with signed surface distance values."""
    _cpp.distance.compute_signed_distance(
        dist_func._cpp_object, soup, cell_map, sign_mode, max_iter, tol
    )


def from_stl(
    mesh: Mesh,
    stl_path: str,
    sign_mode: SignMode = SignMode.ComponentAnchor,
    max_iter: int = 1000,
    tol: float = 1.0e-10,
    aabb_padding: float = 0.05,
) -> Function:
    """Compute a P1 signed distance field from an STL surface."""
    rank = mesh.comm.Get_rank()
    t_start = time.time()

    V = functionspace(mesh, ("Lagrange", 1))
    dist_func = Function(V)
    t_space = time.time()

    soup = distribute_stl(mesh, stl_path, aabb_padding)
    t_dist = time.time()

    cell_map = build_cell_triangle_map(mesh, soup, aabb_padding)
    t_map = time.time()

    compute_signed_distance(dist_func, soup, cell_map, sign_mode, max_iter, tol)
    t_compute = time.time()

    if rank == 0:
        logger = logging.getLogger("cutfemx")
        logger.info("  [Timing] FunctionSpace: %.4fs", t_space - t_start)
        logger.info("  [Timing] Distribute:    %.4fs", t_dist - t_space)
        logger.info("  [Timing] BuildMap:      %.4fs", t_map - t_dist)
        logger.info("  [Timing] FIM+Sign:      %.4fs", t_compute - t_map)

    return dist_func


def reinitialize(
    phi: Function,
    max_iter: int = 200,
    tol: float = 1.0e-10,
) -> None:
    """Reinitialize a level-set function in place as a signed distance field."""
    _cpp.distance.reinitialize(phi._cpp_object, max_iter, tol)


def extend_normal_velocity(
    phi: Function,
    interface_speed: Function,
    interface_cut_mesh,
    *,
    k_ring: int = 1,
    target_space=None,
    normal_sign: float = 1.0,
    max_iter: int = 1000,
    tol: float = 1.0e-10,
) -> NormalExtensionResult:
    """Extend scalar interface speed along normals into the background mesh.

    The C++ carrier solve is first-order and vertex based. If ``target_space``
    is supplied, the carrier fields are interpolated into that scalar Lagrange
    space and the matching vector Lagrange space for the velocity.
    """
    mesh = phi.function_space.mesh
    if interface_cut_mesh.mesh is None:
        raise ValueError("interface_cut_mesh must contain a non-empty mesh")
    if interface_speed.function_space.mesh is not interface_cut_mesh.mesh:
        raise ValueError("interface_speed must live on interface_cut_mesh.mesh")
    if k_ring < 0:
        raise ValueError("k_ring must be non-negative")
    if target_space is not None:
        if target_space.mesh is not mesh:
            raise ValueError("target_space must live on the same mesh as phi")
        if target_space.value_size != 1:
            raise ValueError("target_space must be scalar")

    carrier_space = functionspace(mesh, ("Lagrange", 1))
    carrier_vector_space = _vector_lagrange_space(mesh, 1)
    speed = Function(carrier_space, name="extension_speed")
    velocity = Function(carrier_vector_space, name="extension_velocity")
    signed_distance = Function(carrier_space, name="extension_signed_distance")

    _cpp.distance.extend_normal_velocity(
        phi._cpp_object,
        interface_speed._cpp_object,
        interface_cut_mesh._cpp_object,
        speed._cpp_object,
        velocity._cpp_object,
        signed_distance._cpp_object,
        k_ring,
        normal_sign,
        max_iter,
        tol,
    )

    if target_space is None:
        return NormalExtensionResult(speed, velocity, signed_distance)

    degree = _element_degree(target_space)
    target_vector_space = _vector_lagrange_space(mesh, degree)
    target_speed = Function(target_space, name=speed.name)
    target_velocity = Function(target_vector_space, name=velocity.name)
    target_signed_distance = Function(target_space, name=signed_distance.name)
    target_speed.interpolate(speed)
    target_velocity.interpolate(velocity)
    target_signed_distance.interpolate(signed_distance)
    return NormalExtensionResult(
        target_speed,
        target_velocity,
        target_signed_distance,
    )


def adapt_mesh_to_stl(
    mesh: Mesh,
    stl_path: str,
    *,
    nlevels: int = 1,
    k_ring: int = 1,
    aabb_padding: float = 0.0,
    ghost_mode: GhostMode = GhostMode.shared_facet,
    option: RefinementOption = RefinementOption.parent_cell,
) -> Mesh:
    """Refine a mesh around cells intersected by an STL surface."""
    partitioner = create_cell_partitioner(ghost_mode, 2)

    for _ in range(nlevels):
        edges = _cpp.distance.refinement_edges_from_stl(
            mesh.comm, mesh._cpp_object, stl_path, aabb_padding, k_ring
        )
        if edges.size == 0:
            break
        mesh, _, _ = refine(mesh, edges=edges, partitioner=partitioner, option=option)

    return mesh


compute_distance_from_stl = from_stl
distribute_stl_facets = distribute_stl

File: python/cutfemx/wrappers/distance.cpp
// Copyright (c) 2026 ONERA
// Authors: Susanne Claus
// This file is part of CutFEMx
//
// SPDX-License-Identifier:    MIT

#include <MPICommWrapper.h>
#include <array.h>
#include <caster_mpi.h>

#include <nanobind/nanobind.h>
#include <nanobind/ndarray.h>
#include <nanobind/stl/array.h>
#include <nanobind/stl/pair.h>
#include <nanobind/stl/shared_ptr.h>
#include <nanobind/stl/string.h>
#include <nanobind/stl/vector.h>

#include <algorithm>
#include <cstdint>
#include <memory>
#include <string>
#include <vector>

#include <dolfinx/fem/Function.h>
#include <dolfinx/mesh/Mesh.h>

#include <cutfemx/distance/fast_iterative.h>
#include <cutfemx/distance/normal_extension.h>
#include <cutfemx/distance/reinitialize.h>
#include <cutfemx/distance/sign.h>
#include <cutfemx/distance/stl/cell_triangle_map.h>
#include <cutfemx/distance/stl/distribute.h>
#include <cutfemx/distance/stl/mesh_adapt.h>
#include <cutfemx/distance/stl/reader.h>
#include <cutfemx/distance/stl/surface.h>
#include <cutfemx/mesh/cut_mesh.h>

namespace nb = nanobind;

namespace
{
template <typename Real>
void declare_distance(nb::module_& m, std::string type)
{
  using TriSoup = cutfemx::distance::TriSoup<Real>;
  using CellTriangleMap = cutfemx::distance::CellTriangleMap;

  const std::string trisoup_name = "TriSoup_" + type;
  nb::class_<TriSoup>(m, trisoup_name.c_str(), "Distributed STL triangle soup")
      .def(nb::init<>())
      .def_prop_ro("num_vertices", &TriSoup::nV)
      .def_prop_ro("num_triangles", &TriSoup::nT)
      .def_prop_ro(
          "vertices",
          [](const TriSoup& soup)
          {
            return nb::ndarray<nb::numpy, const Real>(
                soup.X.data(), {static_cast<std::size_t>(soup.nV()), 3ul},
                nb::handle());
          },
          nb::rv_policy::reference_internal)
      .def_prop_ro(
          "triangles",
          [](const TriSoup& soup)
          {
            return nb::ndarray<nb::numpy, const std::int32_t>(
                soup.tri.data(),
                {static_cast<std::size_t>(soup.nT()), 3ul}, nb::handle());
          },
          nb::rv_policy::reference_internal)
      .def("bounding_box",
           [](const TriSoup& soup)
           { return cutfemx::distance::bounding_box(soup); });

  if (type == "float64")
  {
    nb::class_<CellTriangleMap>(m, "CellTriangleMap",
                                "CSR map from local cells to local STL triangles")
        .def(nb::init<>())
        .def_prop_ro("num_cells", &CellTriangleMap::num_cells)
        .def("triangles",
             [](const CellTriangleMap& map, std::int32_t cell)
             {
               auto tris = map.triangles(cell);
               return dolfinx_wrappers::as_nbarray(
                   std::vector<std::int32_t>(tris.begin(), tris.end()));
             });
  }

  m.def("compute_stl_bbox",
        [](const std::string& path)
        { return cutfemx::distance::compute_stl_bbox<Real>(path); },
        nb::arg("path"));

  m.def(
      "distribute_stl",
      [](dolfinx_wrappers::MPICommWrapper comm,
         std::shared_ptr<dolfinx::mesh::Mesh<Real>> mesh,
         const std::string& stl_path, double aabb_padding)
      {
        cutfemx::distance::DistributeSTLOptions opt;
        opt.aabb_padding = aabb_padding;
        return std::make_shared<TriSoup>(
            cutfemx::distance::distribute_stl_facets<Real>(
                comm.get(), *mesh, stl_path, opt));
      },
      nb::arg("comm"), nb::arg("mesh"), nb::arg("stl_path"),
      nb::arg("aabb_padding") = 0.05);

  m.def(
      "build_cell_triangle_map",
      [](std::shared_ptr<const dolfinx::mesh::Mesh<Real>> mesh,
         std::shared_ptr<const TriSoup> soup, double aabb_padding)
      {
        cutfemx::distance::CellTriangleMapOptions opt;
        opt.aabb_padding = aabb_padding;
        return std::make_shared<CellTriangleMap>(
            cutfemx::distance::build_cell_triangle_map<Real>(*mesh, *soup, opt));
      },
      nb::arg("mesh"), nb::arg("soup"), nb::arg("aabb_padding") = 0.0);

  m.def(
      "compute_unsigned_distance",
      [](std::shared_ptr<dolfinx::fem::Function<Real>> dist_func,
         std::shared_ptr<const TriSoup> soup,
         std::shared_ptr<const CellTriangleMap> cell_map, int max_iter,
         double tol)
      {
        cutfemx::distance::FMMOptions opt;
        opt.max_iter = max_iter;
        opt.tol = tol;
        cutfemx::distance::compute_unsigned_distance<Real>(
            *dist_func, *soup, *cell_map, opt);
      },
      nb::arg("dist_func"), nb::arg("soup"), nb::arg("cell_map"),
      nb::arg("max_iter") = 1000, nb::arg("tol") = 1e-10);

  m.def(
      "compute_signed_distance",
      [](std::shared_ptr<dolfinx::fem::Function<Real>> dist_func,
         std::shared_ptr<const TriSoup> soup,
         std::shared_ptr<const CellTriangleMap> cell_map,
         cutfemx::distance::SignMode sign_mode, int max_iter, double tol)
      {
        cutfemx::distance::FMMOptions fmm_opt;
        fmm_opt.max_iter = max_iter;
        fmm_opt.tol = tol;

        std::vector<std::int32_t> closest_tri;
        cutfemx::distance::compute_unsigned_distance<Real>(
            *dist_func, *soup, *cell_map, fmm_opt, &closest_tri);

        cutfemx::distance::SignOptions sign_opt;
        sign_opt.mode = sign_mode;
        cutfemx::distance::apply_sign<Real>(
            *dist_func, *soup, *cell_map, closest_tri, sign_opt);
      },
      nb::arg("dist_func"), nb::arg("soup"), nb::arg("cell_map"),
      nb::arg("sign_mode") = cutfemx::distance::SignMode::ComponentAnchor,
      nb::arg("max_iter") = 1000, nb::arg("tol") = 1e-10);

  m.def(
      "reinitialize",
      [](std::shared_ptr<dolfinx::fem::Function<Real>> phi, int max_iter,
         double tol)
      {
        cutfemx::distance::FMMOptions opt;
        opt.max_iter = max_iter;
        opt.tol = tol;
        cutfemx::distance::reinitialize<Real>(*phi, opt);
      },
      nb::arg("phi"), nb::arg("max_iter") = 1000, nb::arg("tol") = 1e-10);

  m.def(
      "reinitialize_from_facets",
      [](std::shared_ptr<dolfinx::fem::Function<Real>> phi,
         nb::ndarray<const std::int32_t, nb::ndim<1>, nb::c_contig> facets,
         int max_iter, double tol)
      {
        cutfemx::distance::FMMOptions opt;
        opt.max_iter = max_iter;
        opt.tol = tol;
        cutfemx::distance::reinitialize_from_facets<Real>(
            *phi, std::span<const std::int32_t>(facets.data(), facets.size()),
            opt);
      },
      nb::arg("phi"), nb::arg("facets"), nb::arg("max_iter") = 1000,
      nb::arg("tol") = 1e-10);

  m.def(
      "extend_normal_velocity",
      [](std::shared_ptr<const dolfinx::fem::Function<Real>> phi,
         std::shared_ptr<const dolfinx::fem::Function<Real>> interface_speed,
         const cutfemx::mesh::CutMesh<Real>& interface_cut_mesh,
         std::shared_ptr<dolfinx::fem::Function<Real>> speed,
         std::shared_ptr<dolfinx::fem::Function<Real>> velocity,
         std::shared_ptr<dolfinx::fem::Function<Real>> signed_distance,
         int k_ring, double normal_sign, int max_iter, double tol)
      {
        cutfemx::distance::FMMOptions opt;
        opt.max_iter = max_iter;
        opt.tol = tol;
        cutfemx::distance::extend_normal_velocity<Real>(
            std::move(phi), *interface_speed, interface_cut_mesh, *speed,
            *velocity, *signed_distance, k_ring, normal_sign, opt);
      },
      nb::arg("phi"), nb::arg("interface_speed"),
      nb::arg("interface_cut_mesh"), nb::arg("speed"), nb::arg("velocity"),
      nb::arg("signed_distance"), nb::arg("k_ring") = 1,
      nb::arg("normal_sign") = 1.0, nb::arg("max_iter") = 1000,
      nb::arg("tol") = 1.0e-10);

  m.def(
      "refinement_edges_from_stl",
      [](dolfinx_wrappers::MPICommWrapper,
         dolfinx::mesh::Mesh<Real>& mesh, const std::string& stl_path,
         double aabb_padding, int k_ring)
      {
        auto edges = cutfemx::distance::stl::refinement_edges_from_stl(
            mesh, stl_path, aabb_padding, k_ring);

        auto* data = new std::int32_t[edges.size()];
        std::copy(edges.begin(), edges.end(), data);
        nb::capsule owner(data,
                          [](void* p) noexcept
                          { delete[] static_cast<std::int32_t*>(p); });
        return nb::ndarray<std::int32_t, nb::numpy>(
            data, {edges.size()}, owner);
      },
      nb::arg("comm"), nb::arg("mesh"), nb::arg("stl_path"),
      nb::arg("aabb_padding") = 0.0, nb::arg("k_ring") = 1);
}
} // namespace

namespace cutfemx_wrappers
{
void distance(nb::module_& m)
{
  nb::enum_<cutfemx::distance::SignMode>(m, "SignMode")
      .value("LocalNormalBand", cutfemx::distance::SignMode::LocalNormalBand)
      .value("ComponentAnchor", cutfemx::distance::SignMode::ComponentAnchor)
      .value("WindingNumber", cutfemx::distance::SignMode::WindingNumber)
      .export_values();

  declare_distance<float>(m, "float32");
  declare_distance<double>(m, "float64");
}
} // namespace cutfemx_wrappers
